### DISCUSSION TITLE
how to use LogsDao().SaveRequest()

### DISCUSSION THREAD

**Discussion by Oggy4th** [pocketbase/pocketbase#2717] [Q&A] [answered]
Can I use this function like this?

```go
model := &models.Request{
	Url:       e.HttpContext.Request().URL.RequestURI(),
	Method:    strings.ToUpper(e.HttpContext.Request().Method),
	Status:    e.HttpContext.Response().Status,
	Auth:      "test",
	UserIp:    e.HttpContext.RealIP(),
	RemoteIp:  e.HttpContext.Request().RemoteAddr,
	Referer:   e.HttpContext.Request().Referer(),
	UserAgent: e.HttpContext.Request().UserAgent(),
	Meta:      types.JsonMap{"message": "test"},

err := app.LogsDao().SaveRequest(model)
if err != nil {
	fmt.Println(err)
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
Yes, the above should work as long as `app.LogsDao()` is called after app bootstrap.

_Also keep in mind that the logs format will change sometime in the future when introducing "generalized logs" (there are no ETAs for that yet)_

**Reply by Oggy4th** [reply]
Is there any other way to write log?

**Reply by ganigeorgiev** [reply]
What exactly you are trying todo? Please try to provide as much details as possible with your questions since I don't have context about your use case and I'm not sure what to recommend.

Note that we currently don't store logs other than requests. As mentioned above, support for custom logs will be added with the planned logs generalization sometime in the future.

If you just want to attach request logs to a custom route, you can register the `apis.ActivityLogger(app)` middleware to your route.

**Reply by Oggy4th** [reply]
I am calling multiple API requests inside event hooks using go routine. I wanna catch request errors and write some log.

```go
database.App.OnRecordAfterCreateRequest().Add(func(e *core.RecordCreateEvent) error {
	collectionName := e.Record.Collection().Name

	if collectionName == "test" {
		branches, err := BranchModule.FindAllBranches(database.App.Dao())
		if err != nil {
			return apis.NewApiError(http.StatusInternalServerError, "Internal server error", err.Error())
		}

		for _, branch := range branches {
			go func(data map[string]any, publicIp, publicPort string) {
				err := InfoModule.UpdateData(data, publicIp, publicPort)
				if err != nil {
					apis.NewApiError(http.StatusInternalServerError, "Internal server error", err.Error())
				}
			}(e.Record.SchemaData(), branch.PublicIp, branch.PublicPort)
		}
	}
	return nil
})
```

**Reply by ganigeorgiev** [reply]
As mentioned above, at the moment we don't have builtin mechanism for storing general app runtime logs (including errors).

An alternative to your example of using the request log model, could be to create your own `logs` collection (eg. with "data" as `json` field) and submit records to it. The drawback is that you'll have to manually take care for the logs retention (_aka. cleaning up the collection table after some time_) and that the logs will be stored in the main data.db file.
